### ISSUE TITLE
Move the CSLA Serializable attribute to the System namespace

### ISSUE THREAD

**Issue by rockfordlhotka** [MarimerLLC/csla#407]
When the `Csla.Serialization.Serializable` attribute was added to CSLA it was for Silverlight, so probably in 2007 or something.

At that time I didn't want to "pollute" any Microsoft namespaces because I thought that'd cause issues. It turns out that the reverse is true, and that it would have made everyone's life a lot easier to have added `System.Serializable` into the platforms that don't already have it, just like we did when adding the `System.ComponentModel.DataAnnotations` functionality.

I don't think changing this now will cause anyone an issue, because all code using the attribute must already have `using System` statements - except perhaps for the rare case where the code is *only* used in SL/WinRT and doesn't otherwise need anything from the System namespace.

In other words, I think moving the attribute is beneficial going forward, and has a minimal chance of breaking existing code.

**Comment by ChristopherHlusak** [comment]
I like it and you're right under Csla.Serialization it just created compile errors when linking the file from a Silverlight project and having forgotten to add the using that one would have to rectify.

I don't see any problems with it at all.

**Comment by tfreitasleal** [comment]
This a breaking change and there is no point repeating it will break existing code (that's what a breaking change is). Code generators will have to address this breaking change. CslaGenFork will have another option "use Csla.Serialization", but just for Silvelight projects.

**Comment by rockfordlhotka** [comment]
I don't think it will be a breaking change for most people, because nearly all code already includes `using System`.

The `using Csla.Serialization` statements won't *stop* working, they'll just be useless 99.9% of the time now.

This change would only occur in 4.6.200 and higher, which doesn't support Silverlight, so I don't see an issue there either.

**Comment by JasonBock** [comment]
If you do this, remember that the analyzer will need to change so it doesn't generate "using Csla.Serialization;".

**Comment by rockfordlhotka** [comment]
Yes Jason, that's one of my motivations for making this change - and because I really should have done it this way in the first place (hindsight is 20/20).

**Comment by dkehring** [comment]
This is an issue now with .NET Core. Autofac also decided to create a Serializeable attribute within the System namespace and it's a collision with CSLA's Serializable attribute. The reason it's a .NET Core problem is that you can't use "extern alias" to disambiguate references like you could before in the full .NET Framework. Or maybe the problem is poluting System.* namespaces?

**Comment by JasonBock** [comment]
Yeah, this is now affecting a project I'm on, and I don't know of a way to get around it other than by going back a version with either CSLA or Autofac, neither of which I like.

**Comment by rockfordlhotka** [comment]
I suppose the code won't even compile because the two attributes conflict?

**Comment by rockfordlhotka** [comment]
This issue has been raised on the Autofac site too (https://github.com/autofac/Autofac/issues/782) - clearly we're going to need a solution.

**Comment by rockfordlhotka** [comment]
It sounds perhaps, like one solution would be for CSLA to:

1. Externalize the `System.SerializableAttribute` type into its own NuGet package
2. Require that the app provide a reference to some `IsSerializable` provider. In full .NET this provider would rely on normal reflection, if someone is purely using CSLA they could use an implementation included in the NuGet package from step 1, and if someone is also using Autofac they could implement their own `IsSerializable` provider that makes use of the Autofac attribute.

This solution would not require that Autofac do anything at all, though they might also consider following a similar approach so as to provide a workaround for future conflicts.

**Comment by JasonBock** [comment]
I personally wouldn't do anything. In my mind, this is a .NET Core issue and not an issue with either CLSA or Autofac.

Here's some info that may work in terms of doing aliases in .NET Core. I haven't tried them but I'm going to give it a shot in a moment:

https://github.com/dotnet/cli/issues/564

**Comment by dkehring** [comment]
I agree it's a .NET Core issue since they don't support a way to alias references. It must be assumed there will be namespace collisions.

**Comment by rockfordlhotka** [comment]
Autofac has pushed their latest build to NuGet, and it resolves the issue by removing their `Serializable` attribute.

**Comment by dkehring** [comment]
Thanks for the info Rocky!
